### PR TITLE
feat(levm): cache refactor

### PR DESCRIPTION
**Pull request by ilitteri**
**Motivation**

**Description**

- A set `touched_accounts` is introduced to track read accounts.
- Cache is no longer used as a tracker of read accounts. It is now only used to track updated accounts. All the account update logic is encapsulated in individual functions.
- Cache struct is removed and a `cache` module is introduced with the cache's behavior.
- Gas cost functions where the access cost takes place were refactored.

**Leftovers**

- https://github.com/lambdaclass/ethrex/issues/1316

**EF Test Report**

8 more tests pass without any further changes

```sh
✓ Ethereum Foundation Tests: 714 passed 3387 failed 4101 total run - 12:08
✓ Summary: 714/4101

Cancun: 627/3578
Shanghai: 33/221
Homestead: 0/17
Istanbul: 0/34
London: 17/39
Byzantium: 0/33
Berlin: 17/35
Constantinople: 0/66
Merge: 20/62
Frontier: 0/16
```

### CODE REVIEW

File: crates/vm/levm/src/vm.rs [hunk 1]
```diff
@@ -110,6 +126,9 @@ impl VM {
                     accrued_substate: Substate::default(),
                     cache,
                     tx_kind: to,
+                    // FIXME
```
Comment (maximopalopoli) [inline crates/vm/levm/src/vm.rs:129]: What would be the thing that needs to be fixed here? I guess adding the accounts from the upper layer, but a comment or an issue would be better to make sure someone fixes it in the future.

File: crates/vm/levm/src/gas_cost.rs [hunk 2]
```diff
@@ -618,3 +501,227 @@ pub fn tx_creation(code_length: u64, number_of_words: u64) -> Result<u64, OutOfG
...
+
+    // Note: code_execution_cost will be charged from the sub context post-state.
+    let dynamic_gas = memory_expansion_cost
+        .checked_add(address_access_cost)
+        .ok_or(OutOfGasError::GasCostOverflow)?;
+
+    Ok(static_gas
+        .checked_add(dynamic_gas)
+        .ok_or(OutOfGasError::GasCostOverflow)?)
+}
```
Comment (maximopalopoli) [inline crates/vm/levm/src/gas_cost.rs:727]: Maybe the call type opcodes can be abstracted to a call_type_cost, or similar

**Review by maximopalopoli — APPROVED** [review]
LGTM, I just leave you some comments

File: crates/vm/levm/src/vm.rs (hunk 1 above)
Comment (ilitteri) [reply crates/vm/levm/src/vm.rs:129]: Oh, it is a leftover, I'll remove it asap.

Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:129]:
```suggestion
```

File: crates/vm/levm/src/vm.rs [hunk 3]
```diff
@@ -907,53 +851,187 @@ impl VM {
...
+        (account, address_was_cold)
+    }
+
+    /// Accesses to an account's storage slot.
+    ///
+    /// Accessed storage slots are stored in the `touched_storage_slots` set.
+    /// Accessed storage slots take place in some gas cost computation.
+    #[must_use]
+    pub fn access_storage_slot(&mut self, address: Address, key: H256) -> (StorageSlot, bool) {
+        let storage_slot_was_cold = self.touched_storage_slots.insert(key);
```
[the remark this answers is not in the document]
Comment (ilitteri) [reply crates/vm/levm/src/vm.rs]: Fixed in [a3382c7](https://github.com/lambdaclass/ethrex/pull/1314/commits/a3382c752b957e6b3b0ff44f6d70c682b513f544).

File: crates/vm/levm/src/vm.rs (hunk 1 above)
Comment (ilitteri) [reply crates/vm/levm/src/vm.rs]: Done in [729e035](https://github.com/lambdaclass/ethrex/pull/1314/commits/729e035a913b1679ea46a587dc6c5dd587271e7b).

File: crates/vm/levm/src/gas_cost.rs [hunk 4]
```diff
@@ -276,49 +300,52 @@ fn mem_expansion_behavior(
...
     } else {
         U256::from(100)
     };
 
-    base_dynamic_gas
-        .checked_add(sstore_gas_cost)
-        .ok_or(OutOfGasError::GasCostOverflow)
+    if storage_slot_was_cold {
+        base_dynamic_gas = base_dynamic_gas
+            .checked_add(U256::from(2100))
```
Comment (jrchatruc) [inline crates/vm/levm/src/gas_cost.rs:338]: Can't we use constants from `gas_cost.rs` for these and the other gas costs we are adding/substracting?

File: crates/vm/levm/src/vm.rs [hunk 5]
```diff
@@ -907,53 +844,191 @@ impl VM {
...
-                self.cache.add_account(address, &acc);
-                slot
+                cache::insert_account(&mut self.cache, address, account.clone());
+                account
             }
         }
     }
+
+    // /// Gets storage slot, first checking the cache and then the database (caching in the second case)
+    // pub fn get_storage_slot(&mut self, address: Address, key: H256) -> StorageSlot {
```
Comment (jrchatruc) [inline crates/vm/levm/src/vm.rs:1018]: Should we delete this?

Comment (ilitteri) [reply crates/vm/levm/src/vm.rs:1018]: Yes! My bad.

File: crates/vm/levm/src/gas_cost.rs (hunk 4 above)
Comment (ilitteri) [reply crates/vm/levm/src/gas_cost.rs:338]: Yes!

File: crates/vm/levm/src/vm.rs (hunk 5 above)
Comment (ilitteri) [reply crates/vm/levm/src/vm.rs]: Done in [c83a496](https://github.com/lambdaclass/ethrex/pull/1314/commits/c83a4964132f911d081fc4a5ab8bbf71d49480ff)

File: crates/vm/levm/src/gas_cost.rs (hunk 4 above)
Comment (ilitteri) [reply crates/vm/levm/src/gas_cost.rs]: Done in [0835db6](https://github.com/lambdaclass/ethrex/pull/1314/commits/0835db6f1c6bca77cc429ad599e584a2094a4e5b) and [d78436a](https://github.com/lambdaclass/ethrex/pull/1314/commits/d78436a9feead6a263b05283a58cd34eb45f255b)

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**